### ISSUE TITLE
Feature request: allow passing auxiliary data when defining custom gradients

### ISSUE THREAD

**Issue by nkitaev** [jax-ml/jax#844]
Currently, jax allows defining custom gradients (where the gradient of a function `a -> b` has the form `CT b -> CT a`), but does not allow passing through auxiliary variables. Allowing auxiliary data would be particularly useful for reversible networks, where the backward pass has the form `b -> CT b -> (a, CT a)`.

**Comment by wangpengmit** [comment]
It's also useful to support `jax.defvjp_all(f, rule)` where `f` has auxiliary output (e.g. `f` is used as `y, new_state = f(x, params, state)`).

**Comment by dfm** [comment]
I'm going to close this old issue because it is now possible to implement things like this using [`jax.custom_vjp`](https://docs.jax.dev/en/latest/_autosummary/jax.custom_vjp.html). If anyone comes across this thread and has further questions, please feel free to open a new issue with your specific details!
